File: src/software/ai/hl/stp/play/kickoff_enemy/kickoff_enemy_play_test.py
import threading

import software.python_bindings as tbots_cpp
from proto.play_pb2 import PlayName

from software.simulated_tests.validation.robot_enters_region import (
    NumberOfRobotsEventuallyEntersRegion,
    NumberOfRobotsAlwaysStaysInRegion,
    RobotNeverEntersRegion,
)
from software.simulated_tests.validation.or_validation import OrValidation
from software.simulated_tests.validation.ball_enters_region import BallNeverEntersRegion
from proto.message_translation.tbots_protobuf import create_world_state
from proto.import_all_protos import Command
from proto.ssl_gc_common_pb2 import Team
from software.simulated_tests.simulated_test_fixture import (
    pytest_main,
)


def test_kickoff_enemy_play(simulated_test_runner):
    ball_initial_pos = tbots_cpp.Point(0, 0)
    field = tbots_cpp.Field.createSSLDivisionBField()

    def setup(*args):
        blue_bots = [
            tbots_cpp.Point(-3, 2.5),
            tbots_cpp.Point(-2.8, 2.5),
            tbots_cpp.Point(-2.6, 2.5),
            tbots_cpp.Point(-3, -2.5),
            tbots_cpp.Point(-2.8, -2.5),
            tbots_cpp.Point(-2.6, -2.5),
        ]

        yellow_bots = [
            tbots_cpp.Point(2, 0),
            tbots_cpp.Point(2, 2.5),
            tbots_cpp.Point(2, -2.5),
            field.enemyGoalCenter(),
            field.enemyDefenseArea().negXNegYCorner(),
            field.enemyDefenseArea().negXPosYCorner(),
        ]

        simulated_test_runner.set_world_state(
            create_world_state(
                blue_robot_locations=blue_bots,
                yellow_robot_locations=yellow_bots,
                ball_location=ball_initial_pos,
                ball_velocity=tbots_cpp.Vector(0, 0),
            ),
        )

        simulated_test_runner.send_gamecontroller_command(
            gc_command=Command.Type.STOP, team=Team.UNKNOWN
        )
        simulated_test_runner.send_gamecontroller_command(
            gc_command=Command.Type.KICKOFF, team=Team.YELLOW
        )

        # Let robots get ready before starting kickoff
        threading.Timer(
            4.0,
            lambda: simulated_test_runner.send_gamecontroller_command(
                gc_command=Command.Type.NORMAL_START, team=Team.YELLOW
            ),
        ).start()

        simulated_test_runner.set_plays(
            blue_play=PlayName.KickoffEnemyPlay,
            yellow_play=PlayName.KickoffFriendlyPlay,
        )

    # Two friendly robots in position to shadow enemy robots. Rectangles are
    # chosen to be generally in the way of the the front 3 enemy robots and the
    # friendly goal, based on where the enemy robots are initialized in the test.
    shadowing_rect_1 = tbots_cpp.Rectangle(
        tbots_cpp.Point(-0.4, 1.0), tbots_cpp.Point(0, 1.5)
    )
    shadowing_rect_2 = tbots_cpp.Rectangle(
        tbots_cpp.Point(-0.4, -1.5), tbots_cpp.Point(0, -1.0)
    )
    shadowing_rect_3 = tbots_cpp.Rectangle(
        tbots_cpp.Point(-0.86, -0.1), tbots_cpp.Point(-0.60, 0.1)
    )

    # Two Friendly robots defending the exterior of defense box
    robots_defending_rect = tbots_cpp.Rectangle(
        tbots_cpp.Point(-3.5, -1.1), tbots_cpp.Point(-3.2, 1.1)
    )

    eventually_validation_sequence_set = [
        [
            NumberOfRobotsEventuallyEntersRegion(
                regions=[shadowing_rect_1], req_robot_cnt=1
            )
        ],
        [
            NumberOfRobotsEventuallyEntersRegion(
                regions=[shadowing_rect_2], req_robot_cnt=1
            )
        ],
        [
            NumberOfRobotsEventuallyEntersRegion(
                regions=[shadowing_rect_3], req_robot_cnt=1
            )
        ],
        [
            NumberOfRobotsEventuallyEntersRegion(
                regions=[robots_defending_rect], req_robot_cnt=2
            ),
        ],
    ]

    friendly_half = field.friendlyHalf()
    center_circle = field.centerCircle()

    # Validation RoboCup SSL rules: can't enter center circle nor enemy half when ball is not kicked yet
    always_validation_sequence_set = [
        [
            OrValidation(
                [
                    RobotNeverEntersRegion(regions=[center_circle]),
                    BallNeverEntersRegion(
                        regions=[tbots_cpp.Circle(tbots_cpp.Point(0, 0), 0.05)]
                    ),
                ]
            ),
            OrValidation(
                [
                    NumberOfRobotsAlwaysStaysInRegion(
                        regions=[friendly_half], req_robot_cnt=6
                    ),
                    BallNeverEntersRegion(
                        regions=[tbots_cpp.Circle(tbots_cpp.Point(0, 0), 0.05)]
                    ),
                ]
            ),
        ]
    ]

    simulated_test_runner.run_test(
        setup=setup,
        inv_eventually_validation_sequence_set=eventually_validation_sequence_set,
        inv_always_validation_sequence_set=always_validation_sequence_set,
        ag_eventually_validation_sequence_set=eventually_validation_sequence_set,
        ag_always_validation_sequence_set=always_validation_sequence_set,
        test_timeout_s=10,
    )


if __name__ == "__main__":
    pytest_main(__file__)

File: src/software/ai/hl/stp/play/kickoff_enemy/kickoff_enemy_play_test.cpp
#include "software/ai/hl/stp/play/kickoff_enemy/kickoff_enemy_play.h"

#include <gtest/gtest.h>

#include "software/geom/algorithms/calculate_block_cone.h"
#include "software/geom/algorithms/contains.h"
#include "software/simulated_tests/cpp_validation/validation_function.h"
#include "software/simulated_tests/non_terminating_validation_functions/robots_in_friendly_half_validation.h"
#include "software/simulated_tests/non_terminating_validation_functions/robots_not_in_center_circle_validation.h"
#include "software/simulated_tests/simulated_er_force_sim_play_test_fixture.h"
#include "software/simulated_tests/terminating_validation_functions/robot_in_polygon_validation.h"
#include "software/test_util/test_util.h"
#include "software/time/duration.h"
#include "software/world/world.h"

class KickoffEnemyPlayTest : public SimulatedErForceSimPlayTestFixture
{
   protected:
    TbotsProto::FieldType field_type = TbotsProto::FieldType::DIV_B;
    Field field                      = Field::createField(field_type);
};

// TODO (#3105): Re-enable test once destinations are moved outside of obstacles
TEST_F(KickoffEnemyPlayTest, DISABLED_test_kickoff_enemy_play)
{
    BallState ball_state(Point(0, 0), Vector(0, 0));
    auto friendly_robots = TestUtil::createStationaryRobotStatesWithId(
        {Point(-field.xLength() / 3.0, 5.0 * field.yLength() / 12.0),
         Point(-field.xLength() / 3.0, field.yLength() / 4.0),
         Point(-field.xLength() / 3.0, field.yLength() / 12.0),
         Point(-field.xLength() / 3.0, -field.yLength() / 12.0),
         Point(-field.xLength() / 3.0, -field.yLength() / 4.0),
         Point(-field.xLength() / 3.0, -5.0 * field.yLength() / 12.0)});
    setFriendlyGoalie(0);
    auto enemy_robots = TestUtil::createStationaryRobotStatesWithId(
        {Point(field.xLength() / 9.0, 0),
         Point(field.xLength() / 9.0, 5.0 * field.yLength() / 12.0),
         Point(field.xLength() / 9.0, -5.0 * field.yLength() / 12.0),
         field.enemyGoalCenter(), field.enemyDefenseArea().negXNegYCorner(),
         field.enemyDefenseArea().negXPosYCorner()});
    setEnemyGoalie(0);
    setAiPlay(TbotsProto::PlayName::KickoffEnemyPlay);
    setRefereeCommand(RefereeCommand::NORMAL_START, RefereeCommand::PREPARE_KICKOFF_THEM);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield)
        {
            const Field test_field = world_ptr->field();
            // Two friendly robots in position to shadow enemy robots. Rectangles are
            // chosen to be generally in the way of the the front 3 enemy robots and the
            // friendly goal, based on where the enemy robots are initialized in the test.
            Rectangle shadowing_rect_1(
                Point(0.0, 1.5 * test_field.yLength() / 6.0),
                Point(-0.4 * test_field.xLength() / 9.0,
                      1.3 * test_field.yLength() / 6.0));
            Rectangle shadowing_rect_2(
                Point(0.0, -1.5 * test_field.yLength() / 6.0),
                Point(-0.4 * test_field.xLength() / 9.0,
                      -1.3 * test_field.yLength() / 6.0));
            Rectangle shadowing_rect_3(
                Point(-0.60 * test_field.xLength() / 9.0,
                      0.1 * test_field.yLength() / 6.0),
                Point(-0.86 * test_field.xLength() / 9.0,
                      -0.1 * test_field.yLength() / 6.0));
            robotInPolygon(shadowing_rect_1, 1, world_ptr, yield);
            robotInPolygon(shadowing_rect_2, 1, world_ptr, yield);
            robotInPolygon(shadowing_rect_3, 1, world_ptr, yield);

            // Two Friendly robots defending the exterior of defense box
            Rectangle robotsDefendingRect(
                Point(test_field.friendlyDefenseArea().xMax() +
                          0.3 * test_field.xLength() / 9.0,
                      1.1 * test_field.yLength() / 6.0),
                Point(test_field.friendlyDefenseArea().xMax(),
                      -1.1 * test_field.yLength() / 6.0));
            robotInPolygon(robotsDefendingRect, 2, world_ptr, yield);
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {
        robotsInFriendlyHalf, robotsNotInCenterCircle};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}
